server/entity/experience.go

```diff
@@ -0,0 +1,193 @@
+package entity
+
+import (
+	"math"
+	"sync"
+)
+
+var maxLevel int32 = math.MaxInt32
+
+type ExperienceManager struct {
+	mu              sync.RWMutex
+	totalExperience int
+}
+
+// NewExperienceManager return a ExperienceManager with default value of level, progress and totalExperience that are equal to 0.
+func NewExperienceManager() *ExperienceManager {
+	return &ExperienceManager{}
+}
+
+// Level get experience level.
+func (e *ExperienceManager) Level() int32 {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	experience := e.totalExperience
+	var level int32 = 1
+	nextLevel := true
+	for nextLevel {
+		xp := e.ExperienceForLevel(level)
+		if xp <= experience {
+			experience -= xp
+			level++
+		} else {
+			nextLevel = false
+		}
+	}
```

Comment:
This code is duplicated everywhere. Might want to pull this out in its own method?